### PR TITLE
Be able to fund transactions with peg-ins specified

### PR DESCRIPTION
**Pull request by achow101**
Allows `fundrawtransaction` and `walletcreatefundedpsbt` to take transactions with pre-selected peg-in inputs and take them into account when selecting additional inputs to meet the output amounts.

As a side effect, those RPCs will also now allow pre-selected non-wallet inputs so long as the correct additional solving data is provided (scripts and pubkeys are needed to estimate the size for transaction fees).

Based on #751 

Ports https://github.com/bitcoin/bitcoin/pull/17211 with some modification to work with elements outputs and such.

### CODE REVIEW

**Comment by instagibbs** [comment]
awesome work, concept ACK. I should review the the parent PR first.

**Comment by instagibbs** [comment]
could i get a clean rebase on master?

**Comment by achow101** [comment]
Rebased

File: test/functional/feature_fedpeg.py [hunk 1]
```diff
@@ -317,6 +317,13 @@ def run_test(self):
         fin_psbt = sidechain.finalizepsbt(signed_psbt['psbt'])
         assert_equal(fin_psbt, signed_pegin)
 
+        # Try funding a psbt with the peg-in
```
Comment (instagibbs) [inline test/functional/feature_fedpeg.py]: Assert that the balance is 50 BTC and the outputs summed to too much for the value you're adding next line.

File: test/functional/rpc_fundrawtransaction.py [hunk 2]
```diff
@@ -777,5 +777,26 @@ def run_test(self):
...
+        raw_tx = self.nodes[2].createrawtransaction([{"txid": txid, "vout": vout}], {addr_info['unconfidential']: 20})
+        assert_raises_rpc_error(-4, "Missing solving data for estimating transaction size", self.nodes[2].fundrawtransaction, raw_tx)
+
+        # But funding should work when the solving data is provided
+        funded_tx = self.nodes[2].fundrawtransaction(raw_tx, {}, False, {"pubkeys": [addr_info['pubkey']]})
+        signed_tx = self.nodes[2].signrawtransactionwithwallet(funded_tx['hex'])
+        assert not signed_tx['complete']
+        signed_tx = self.nodes[0].signrawtransactionwithwallet(signed_tx['hex'])
+        assert signed_tx['complete']
+        # Don't send because we didn't blind it so it's not actually valid.
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/rpc_fundrawtransaction.py]: `blindrawtransaction` didn't like non-wallet inputs.

File: src/wallet/coincontrol.h [hunk 3]
```diff
@@ -55,11 +57,32 @@ class CCoinControl
...
+        txout = ext_it->second;
+        return true;
+    }
+
     void Select(const COutPoint& output)
     {
         setSelected.insert(output);
     }
 
+    void Select(const COutPoint& outpoint, const CTxOut& txout)
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/coincontrol.h]: Done

File: src/wallet/coincontrol.h [hunk 4]
```diff
@@ -83,6 +83,16 @@ class CCoinControl
         m_external_txouts.emplace(outpoint, txout);
     }
 
+    void Select(const COutPoint& outpoint, const Sidechain::Bitcoin::CTxOut& txout_in)
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/coincontrol.h]: Done

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -3381,10 +3381,17 @@ void FundTransaction(CWallet* const pwallet, CMutableTransaction& tx, CAmount& f
...
+    // Check any existing inputs for peg-in data and add to external txouts if so
     // Fetch specified UTXOs from the UTXO set
     std::map<COutPoint, Coin> coins;
-    for (const CTxIn& txin : tx.vin) {
+    for (unsigned int i = 0; i < tx.vin.size(); ++i ) {
+        const CTxIn& txin = tx.vin[i];
         coins[txin.prevout]; // Create empty map entry keyed by prevout.
+        if (txin.m_is_pegin) {
+            CScriptWitness& pegin_witness = tx.witness.vtxinwit[i].m_pegin_witness;
+            CTxOut txout = GetPeginOutputFromWitness(pegin_witness);
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/rpcwallet.cpp]: Done

File: test/functional/feature_fedpeg.py (hunk 1 above)
Comment (achow101) [reply test/functional/feature_fedpeg.py]: Done

File: test/functional/rpc_fundrawtransaction.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/rpc_fundrawtransaction.py]: Future Work: Enable `blindrawtransaction` to know that a bitcoin-style peg-in input has a null blinding factor.

**Comment by instagibbs** [comment]
utACK https://github.com/ElementsProject/elements/pull/755/commits/00eb6c0978c8cb42b5da68381d8871159fe9b5af

File: test/functional/rpc_fundrawtransaction.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/rpc_fundrawtransaction.py]: This test is just for external inputs in general, not peg-ins.

**Pull request merged**